In [1]:
import pandas as pd
import numpy as np
# Data Engineering\Data\

In [2]:
df = pd.read_csv('D:\Repositorios\Practicas\Data Engineering\Data\winemag-data-130k-v2.csv')

In [3]:
df.shape

(129971, 14)

In [4]:
df.columns

Index(['Unnamed: 0', 'country', 'description', 'designation', 'points',
       'price', 'province', 'region_1', 'region_2', 'taster_name',
       'taster_twitter_handle', 'title', 'variety', 'winery'],
      dtype='object')

In [5]:
df_one = df.iloc[:,0:7]
df_two = df.iloc[:,7:14]

In [6]:
df_one.shape

(129971, 7)

In [7]:
df_two.shape

(129971, 7)

In [8]:
frames = [df_one, df_two]

In [9]:
df_complete = pd.concat(frames, sort = False, axis = 1, join = 'outer')

In [10]:
df_complete.shape

(129971, 14)

In [11]:
df_complete.columns

Index(['Unnamed: 0', 'country', 'description', 'designation', 'points',
       'price', 'province', 'region_1', 'region_2', 'taster_name',
       'taster_twitter_handle', 'title', 'variety', 'winery'],
      dtype='object')

# Defino el objetivo de mi análisis:

Quiero determinar las variedades de vinos que tengan dispersión más grande en calificaciones y cuáles son los 
países en los que se dan estos casos

Entrada: Tabla 129971x3: 129971 entradas, columnas: 'variedad', 'país', 'puntos'

Salida: Tabla 10x4: 10 entradas, columnas: 'variedad', 'país', 'promedio', 'dispersión'

In [12]:
df_filtered = df_complete[['country', 'variety', 'points']]

In [13]:
df_filtered.shape

(129971, 3)

In [14]:
df_isNull=pd.DataFrame(data=df_filtered.isnull().sum(), columns=["isNull"])

In [15]:
df_isNull['Percentage'] = (df_filtered.isnull().sum()/df_filtered.shape[0])*100

In [16]:
df_isNull[df_isNull.isNull>0] 

isNull  Percentage
country      63    0.048472
variety       1    0.000769

In [17]:
# Por simplicidad, se ha decidido eliminar las instancias donde uno o más de los parámetros es nulo (ya que es un porcentaje
# inferior al 1 %)

In [18]:
df_filtered_wo_Nulls = df_filtered.dropna()

In [19]:
df_filtered_wo_Nulls.shape

(129907, 3)

In [20]:
df_filtered_groups = df_filtered_wo_Nulls.groupby(['country','variety']).agg({'points': ['count', 'median', 'std']}).reset_index()

In [21]:
df_filtered_groups

country                            variety points                 
                                                    count median       std
0     Argentina                            Barbera      1   85.0       NaN
1     Argentina                            Bonarda    105   87.0  2.587410
2     Argentina           Bordeaux-style Red Blend     89   91.0  2.990750
3     Argentina         Bordeaux-style White Blend      1   83.0       NaN
4     Argentina                     Cabernet Blend      8   88.0  2.549510
5     Argentina                     Cabernet Franc     64   90.0  2.448224
6     Argentina  Cabernet Franc-Cabernet Sauvignon      3   87.0  4.725816
7     Argentina              Cabernet Franc-Malbec      4   91.0  0.000000
8     Argentina                 Cabernet Sauvignon    540   86.0  2.777710
9     Argentina  Cabernet Sauvignon-Cabernet Franc      1   84.0       NaN
10    Argentina          Cabernet Sauvignon-Malbec     12   87.0  5.517877
11    Argentina          Cabernet Sauvignon-Merlot      5   87.0  1.581139
12    Argentina          Cabernet Sauvignon-Shiraz      3   85.0  2.516611
13    Argentina           Cabernet Sauvignon-Syrah      2   86.0  1.414214
14    Argentina                    Cabernet-Malbec      1   86.0       NaN
15    Argentina                     Cabernet-Syrah      1   88.0       NaN
16    Argentina                    Champagne Blend      8   86.0  2.900123
17    Argentina                         Chardonnay    295   85.0  2.361203
18    Argentina               Chardonnay-Sauvignon      2   84.5  0.707107
19    Argentina                Chardonnay-Semillon      1   81.0       NaN
20    Argentina                Chardonnay-Viognier      6   85.0  1.471960
21    Argentina                       Chenin Blanc      2   84.0  0.000000
22    Argentina            Chenin Blanc-Chardonnay      5   86.0  0.836660
23    Argentina                     Gewürztraminer      1   84.0       NaN
24    Argentina                             Malbec   1510   87.0  3.181445
25    Argentina                       Malbec Blend      2   90.5  2.121320
26    Argentina                     Malbec-Bonarda      5   86.0  3.577709
27    Argentina                    Malbec-Cabernet      3   93.0  2.309401
28    Argentina              Malbec-Cabernet Franc     10   88.5  2.830391
29    Argentina          Malbec-Cabernet Sauvignon     56   89.0  3.285545
...         ...                                ...    ...    ...       ...
1582         US                 Viognier-Valdiguié      1   84.0       NaN
1583         US                        White Blend    640   87.0  2.895951
1584         US                         White Port      1   92.0       NaN
1585         US                     White Riesling     12   88.0  3.553701
1586         US                          Zinfandel   2711   88.0  3.112639
1587         US                           Zweigelt      7   88.0  1.573592
1588    Ukraine                           Bastardo      1   85.0       NaN
1589    Ukraine                 Cabernet Sauvignon      1   84.0       NaN
1590    Ukraine                         Chardonnay      1   85.0       NaN
1591    Ukraine                             Merlot      2   84.0  0.000000
1592    Ukraine                     Muscat Hamburg      1   86.0       NaN
1593    Ukraine                           Riesling      1   88.0       NaN
1594    Ukraine                               Rosé      1   84.0       NaN
1595    Ukraine                    Sparkling Blend      6   83.0  0.752773
1596    Uruguay                           Albariño      6   86.5  2.581989
1597    Uruguay                     Cabernet Franc      2   85.0  4.242641
1598    Uruguay                 Cabernet Sauvignon      2   87.0  0.000000
1599    Uruguay                         Chardonnay      3   81.0  1.000000
1600    Uruguay                             Malbec      2   87.0  0.000000
1601    Uruguay                             Merlot      1   84.0       NaN
1602    Uruguay                  

##### Debido a que la analítica se centra en la desviación estándar, aquellas instancias en las que solo exista una puntuación, se debe proceder a eliminarlas:

In [22]:
df_filtered_groups = df_filtered_groups.T.reset_index(drop=True).T

In [23]:
df_filtered_groups.columns = ['country', 'variety', 'count', 'median', 'std']

In [24]:
df_filtered_groups

country                            variety count median       std
0     Argentina                            Barbera     1     85       NaN
1     Argentina                            Bonarda   105     87   2.58741
2     Argentina           Bordeaux-style Red Blend    89     91   2.99075
3     Argentina         Bordeaux-style White Blend     1     83       NaN
4     Argentina                     Cabernet Blend     8     88   2.54951
5     Argentina                     Cabernet Franc    64     90   2.44822
6     Argentina  Cabernet Franc-Cabernet Sauvignon     3     87   4.72582
7     Argentina              Cabernet Franc-Malbec     4     91         0
8     Argentina                 Cabernet Sauvignon   540     86   2.77771
9     Argentina  Cabernet Sauvignon-Cabernet Franc     1     84       NaN
10    Argentina          Cabernet Sauvignon-Malbec    12     87   5.51788
11    Argentina          Cabernet Sauvignon-Merlot     5     87   1.58114
12    Argentina          Cabernet Sauvignon-Shiraz     3     85   2.51661
13    Argentina           Cabernet Sauvignon-Syrah     2     86   1.41421
14    Argentina                    Cabernet-Malbec     1     86       NaN
15    Argentina                     Cabernet-Syrah     1     88       NaN
16    Argentina                    Champagne Blend     8     86   2.90012
17    Argentina                         Chardonnay   295     85    2.3612
18    Argentina               Chardonnay-Sauvignon     2   84.5  0.707107
19    Argentina                Chardonnay-Semillon     1     81       NaN
20    Argentina                Chardonnay-Viognier     6     85   1.47196
21    Argentina                       Chenin Blanc     2     84         0
22    Argentina            Chenin Blanc-Chardonnay     5     86   0.83666
23    Argentina                     Gewürztraminer     1     84       NaN
24    Argentina                             Malbec  1510     87   3.18145
25    Argentina                       Malbec Blend     2   90.5   2.12132
26    Argentina                     Malbec-Bonarda     5     86   3.57771
27    Argentina                    Malbec-Cabernet     3     93    2.3094
28    Argentina              Malbec-Cabernet Franc    10   88.5   2.83039
29    Argentina          Malbec-Cabernet Sauvignon    56     89   3.28554
...         ...                                ...   ...    ...       ...
1582         US                 Viognier-Valdiguié     1     84       NaN
1583         US                        White Blend   640     87   2.89595
1584         US                         White Port     1     92       NaN
1585         US                     White Riesling    12     88    3.5537
1586         US                          Zinfandel  2711     88   3.11264
1587         US                           Zweigelt     7     88   1.57359
1588    Ukraine                           Bastardo     1     85       NaN
1589    Ukraine                 Cabernet Sauvignon     1     84       NaN
1590    Ukraine                         Chardonnay     1     85       NaN
1591    Ukraine                             Merlot     2     84         0
1592    Ukraine                     Muscat Hamburg     1     86       NaN
1593    Ukraine                           Riesling     1     88       NaN
1594    Ukraine                               Rosé     1     84       NaN
1595    Ukraine                    Sparkling Blend     6     83  0.752773
1596    Uruguay                           Albariño     6   86.5   2.58199
1597    Uruguay                     Cabernet Franc     2     85   4.24264
1598    Uruguay                 Cabernet Sauvignon     2     87         0
1599    Uruguay                         Chardonnay     3     81         1
1600    Uruguay                             Malbec     2     87         0
1601    Uruguay                             Merlot     1     84       NaN
1602    Uruguay                         Pinot Noir     5     84    3.1305
1603    Uruguay                          Red Blend    20     87   2.35919
1604    Urug

In [25]:
df_filtered_groups = df_filtered_groups[df_filtered_groups['count'] > 1]

In [26]:
df_filtered_groups

country                            variety count median       std
1     Argentina                            Bonarda   105     87   2.58741
2     Argentina           Bordeaux-style Red Blend    89     91   2.99075
4     Argentina                     Cabernet Blend     8     88   2.54951
5     Argentina                     Cabernet Franc    64     90   2.44822
6     Argentina  Cabernet Franc-Cabernet Sauvignon     3     87   4.72582
7     Argentina              Cabernet Franc-Malbec     4     91         0
8     Argentina                 Cabernet Sauvignon   540     86   2.77771
10    Argentina          Cabernet Sauvignon-Malbec    12     87   5.51788
11    Argentina          Cabernet Sauvignon-Merlot     5     87   1.58114
12    Argentina          Cabernet Sauvignon-Shiraz     3     85   2.51661
13    Argentina           Cabernet Sauvignon-Syrah     2     86   1.41421
16    Argentina                    Champagne Blend     8     86   2.90012
17    Argentina                         Chardonnay   295     85    2.3612
18    Argentina               Chardonnay-Sauvignon     2   84.5  0.707107
20    Argentina                Chardonnay-Viognier     6     85   1.47196
21    Argentina                       Chenin Blanc     2     84         0
22    Argentina            Chenin Blanc-Chardonnay     5     86   0.83666
24    Argentina                             Malbec  1510     87   3.18145
25    Argentina                       Malbec Blend     2   90.5   2.12132
26    Argentina                     Malbec-Bonarda     5     86   3.57771
27    Argentina                    Malbec-Cabernet     3     93    2.3094
28    Argentina              Malbec-Cabernet Franc    10   88.5   2.83039
29    Argentina          Malbec-Cabernet Sauvignon    56     89   3.28554
30    Argentina                      Malbec-Merlot     7     86   2.75162
31    Argentina                Malbec-Petit Verdot     4     91         0
32    Argentina                       Malbec-Syrah    16   85.5   1.78419
33    Argentina                      Malbec-Tannat     3     84   2.08167
34    Argentina                 Malbec-Tempranillo     2   86.5  0.707107
35    Argentina                             Merlot    59     85    2.4138
38    Argentina                            Moscato    11     82   1.77866
...         ...                                ...   ...    ...       ...
1568         US                          Valdiguié     6   86.5   1.54919
1570         US                            Verdejo     4   86.5   3.59398
1571         US                           Verdelho    24     87   2.01983
1572         US                         Vermentino    46   87.5   2.21294
1573         US                              Vidal     3     85    1.1547
1574         US                        Vidal Blanc    40     86   2.03668
1575         US                           Vignoles    11     86   2.20743
1576         US                           Viognier   790     88   2.79468
1577         US                Viognier-Chardonnay     2     86   2.82843
1578         US            Viognier-Gewürztraminer     2     86         0
1579         US            Viognier-Grenache Blanc     2     91         0
1580         US                  Viognier-Marsanne     2     88   2.82843
1581         US                 Viognier-Roussanne     7     89    4.6188
1583         US                        White Blend   640     87   2.89595
1585         US                     White Riesling    12     88    3.5537
1586         US                          Zinfandel  2711     88   3.11264
1587         US                           Zweigelt     7     88   1.57359
1591    Ukraine                             Merlot     2     84         0
1595    Ukraine                    Sparkling Blend     6     83  0.752773
1596    Uruguay                           Albariño     6   86.5   2.58199
1597    Uruguay                     Cabernet Franc     2     85   4.24264
1598    Uruguay                 Cabernet Sauvignon     2     87         0
1599    Urug

In [27]:
df_filtered_groups['std']

1        2.58741
2        2.99075
4        2.54951
5        2.44822
6        4.72582
7              0
8        2.77771
10       5.51788
11       1.58114
12       2.51661
13       1.41421
16       2.90012
17        2.3612
18      0.707107
20       1.47196
21             0
22       0.83666
24       3.18145
25       2.12132
26       3.57771
27        2.3094
28       2.83039
29       3.28554
30       2.75162
31             0
32       1.78419
33       2.08167
34      0.707107
35        2.4138
38       1.77866
          ...   
1568     1.54919
1570     3.59398
1571     2.01983
1572     2.21294
1573      1.1547
1574     2.03668
1575     2.20743
1576     2.79468
1577     2.82843
1578           0
1579           0
1580     2.82843
1581      4.6188
1583     2.89595
1585      3.5537
1586     3.11264
1587     1.57359
1591           0
1595    0.752773
1596     2.58199
1597     4.24264
1598           0
1599           1
1600           0
1602      3.1305
1603     2.35919
1604     0.57735
1606     2.701

In [28]:
df_filtered_groups['std'] = df_filtered_groups['std'].apply(pd.to_numeric)

C:\Users\esteb\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [29]:
df_filtered_groups['std']

1       2.587410
2       2.990750
4       2.549510
5       2.448224
6       4.725816
7       0.000000
8       2.777710
10      5.517877
11      1.581139
12      2.516611
13      1.414214
16      2.900123
17      2.361203
18      0.707107
20      1.471960
21      0.000000
22      0.836660
24      3.181445
25      2.121320
26      3.577709
27      2.309401
28      2.830391
29      3.285545
30      2.751623
31      0.000000
32      1.784190
33      2.081666
34      0.707107
35      2.413797
38      1.778661
          ...   
1568    1.549193
1570    3.593976
1571    2.019829
1572    2.212940
1573    1.154701
1574    2.036683
1575    2.207425
1576    2.794679
1577    2.828427
1578    0.000000
1579    0.000000
1580    2.828427
1581    4.618802
1583    2.895951
1585    3.553701
1586    3.112639
1587    1.573592
1591    0.000000
1595    0.752773
1596    2.581989
1597    4.242641
1598    0.000000
1599    1.000000
1600    0.000000
1602    3.130495
1603    2.359193
1604    0.577350
1606    2.7013

In [45]:
df_top = df_filtered_groups.sort_values('std',ascending = False).head(10)

In [46]:
df_top

country                            variety count median       std
1284      Spain                     Tinta del Toro     3     88  5.686241
1510         US                 Roussanne-Marsanne     2     86  5.656854
10    Argentina          Cabernet Sauvignon-Malbec    12     87  5.517877
1566         US                     Trousseau Gris     3     93  5.196152
1357         US                   Cabernet Pfeffer     2   87.5  4.949747
489      France                             Shiraz     2   87.5  4.949747
1455         US                    Mourvèdre-Syrah     2   90.5  4.949747
6     Argentina  Cabernet Franc-Cabernet Sauvignon     3     87  4.725816
1581         US                 Viognier-Roussanne     7     89  4.618802
417      France                           Grolleau     3     90  4.618802

### Actividades sucesivas: Con el fin de mejorar los resultados obtenidos, se aconseja el ponderar los desvíos con otro criterio; por ejemplo, se puede incorporar un "coeficiente de interés/importancia" que le de mayor valor a aquellas desviaciones basadas en mayor tamaño de muestras 